scaaml/stats/attack_points/aes_128/attack_points.py

```diff
@@ -0,0 +1,334 @@
+                           ciphertext: npt.NDArray[np.uint8], guess: int,
+                           byte_index: int) -> int:
+        assert 0 <= guess < 256
+        return int(SBOX[guess ^ plaintext[byte_index]])
+
+    @staticmethod
+    def target_secret(key: npt.NDArray[np.uint8],
+                      plaintext: npt.NDArray[np.uint8], byte_index: int) -> int:
+        return int(key[byte_index])
+
+
+class LastRoundStateDiff(AttackPointAES128):
+    """Difference of last round states is leaking. Useful for instance for
+    hardware AES (see `https://github.com/newaetech/chipwhisperer-jupyter`
+    `courses/sca201/Lab 2_2 - CPA on Hardware AES Implementation.ipynb`)
+    """
+
+    @staticmethod
+    def leakage_knowing_secrets(key: npt.NDArray[np.uint8],
+                                plaintext: npt.NDArray[np.uint8],
+                                byte_index: int) -> int:
+        invshift_undo = [0, 5, 10, 15, 4, 9, 14, 3, 8, 13, 2, 7, 12, 1, 6, 11]
+        ciphertext = AESSBOX.ciphertext(
+            key=bytearray(key),
+            plaintext=bytearray(plaintext),
+        )
+
+        last_key_schedule = key_schedule(key)
+        correct_k = last_key_schedule[-4:].reshape(-1)
+        guess = correct_k[byte_index]
+
+        st10 = ciphertext[invshift_undo[byte_index]]
+        st9 = SBOX_INV[ciphertext[byte_index] ^ guess]
+        byte_value = st9 ^ st10
+
+        return int(byte_value)
+
+    @staticmethod
+    def leakage_from_guess(plaintext: npt.NDArray[np.uint8],
+                           ciphertext: npt.NDArray[np.uint8], guess: int,
+                           byte_index: int) -> int:
+        assert 0 <= guess < 256
+
+        invshift_undo = [0, 5, 10, 15, 4, 9, 14, 3, 8, 13, 2, 7, 12, 1, 6, 11]
+
+        st10 = ciphertext[invshift_undo[byte_index]]
+        st9 = SBOX_INV[ciphertext[byte_index] ^ guess]
+        byte_value = st9 ^ st10
+
+        return int(byte_value)
+
+    @staticmethod
+    def target_secret(key: npt.NDArray[np.uint8],
+                      plaintext: npt.NDArray[np.uint8], byte_index: int) -> int:
+        last_key_schedule = key_schedule(key)
+        correct_k = last_key_schedule[-4:].reshape(-1)
+        guess = correct_k[byte_index]
+
+        return int(guess)
+
+
+class LeakageModelAES128:
+    """The value that might be correlated with value(s) in the trace. Keeps the
+    byte index and if we should use just the Hamming weight.
+    """
+
+    def __init__(self,
+                 byte_index: int,
+                 attack_point: AttackPointAES128,
+                 use_hamming_weight: bool = True) -> None:
+        """Gives the leakage function.
+
+        Args:
+
+          byte_index (int): Which byte to target (in range(16)).
+
+          attack_point (AttackPointAES128): Use either input or output of the
+          first SBOX.
+
+          use_hamming_weight (bool): Use just the Hamming weight of the value.
+        """
+        assert byte_index in range(16)
```

Comment:
nit: for consistency with the rest fo the file, `0 <= byte_index < 16` would be better